### DISCUSSION TITLE
Injecting dependencies in SwiftUI preview

### DISCUSSION THREAD

**Discussion by JanC** [pointfreeco/swift-composable-architecture#3332] [General]
hey there,
I'm having troubles to inject a dependency for a  SwiftUI preview.  I created a dependency with `liveValue` and `previewValue` and that works as expected

```swift
private enum APIClientKey: DependencyKey {
    static let liveValue: APIClient = RealClient()
    static let previewValue: APIClient = MyMockClient(result: "preview default value")
}
```

Then I'd like to preview a specific use case where I configure the `MyMockClient` dependency so I create my preview like this:

```swift
#Preview {
    NavigationStack {
        MySampleView(
            store: Store(
                initialState: MySampleFeature.State(),
                reducer: { MySampleFeature() },
                withDependencies: {
                    $0.apiClient = MyMockClient(result: "mocked result")
                }
            )
        )
    }
}
```

However, the preview still uses the instance of my mock created in `static let previewValue: APIClient = MyMockClient(result: "preview default value")`.

I put some logs in the initialiser of `MyMockClient` and when I run the preview, I see it's created 2x when I run the SwiftUI preview:

```
MyMockClient.init mocked result
MyMockClient.init preview default value
```

The complete code to reproduce it:

```swift
import ComposableArchitecture
import Foundation
import SwiftUI

protocol APIClient {
    func fetch() -> String
}

class MyMockClient: APIClient {
    let result: String

    init(result: String = "live") {
        print("MyMockClient.init \(result)")
        self.result = result
    }

    func fetch() -> String {
        result
    }
}

private enum APIClientKey: DependencyKey {
    static let liveValue: APIClient = MyMockClient(result: "live value")
    static let previewValue: APIClient = MyMockClient(result: "preview default value")
}

extension DependencyValues {
    var apiClient: APIClient {
        get { self[APIClientKey.self] }
        set { self[APIClientKey.self] = newValue }
    }
}

@Reducer
struct MySampleFeature {
    @Dependency(\.apiClient) var apiClient

    @ObservableState
    struct State: Equatable {
        var title: String = ""
    }

    enum Action {
        case viewDidAppear
    }

    var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case .viewDidAppear:
                state.title = apiClient.fetch()
                return .none
            }
        }
    }
}

struct MySampleView: View {
    @Bindable
    var store: StoreOf<MySampleFeature>

    var body: some View {
        List {
            Text("Hello \(store.title)")
        }

        .onAppear(perform: {
            store.send(.viewDidAppear)
        })
        .navigationTitle("Sample")
    }
}

#Preview {
    NavigationStack {
        MySampleView(
            store: Store(
                initialState: MySampleFeature.State(),
                reducer: { MySampleFeature() },
                withDependencies: {
                    $0.apiClient = MyMockClient(result: "mocked result")
                }
            )
        )
    }
}

```

I also tried with the alternative syntax but the result is the same 

```swift
#Preview {
    NavigationStack {
        MySampleView(
            store: Store(
                initialState: MySampleFeature.State(),
                reducer: {
                    MySampleFeature()
                        .dependency(\.apiClient, MyMockClient(result: "mocked result"))
                }
            )
        )
    }
}
```

What did I miss?

thanks in advance

**Comment by mbrandonw** [comment]
Hi @JanC, when I run your code in a preview it works as I expect:

<img width="1415" alt="image" src="https://github.com/user-attachments/assets/41aa7f29-5835-4abd-8055-7e6dcf8e7db1">

The preview says "Hello mocked result". If you are seeing different can you provide a project that reproduces the problem?

**Reply by JanC** [reply]
So I also noticed that my described behavior sometimes works as expected. I'm trying find out what could be the reason.

**Reply by mbrandonw** [reply]
I'm sorry but I'm never able to ever produce the problem. I have run the preview many times and I have restarted Xcode, as well as tried Xcode 15.4 and Xcode 16 beta 6. Is there anything else special about your set up?

**Reply by JanC** [reply]
I think it could be related to Xcode 16 beta:

![Screenshot 2024-08-30 at 18 45 17](https://github.com/user-attachments/assets/eb1ef59e-8262-4d27-88c0-ed8a4065f8fe)



Xcode 15.3.x seems to work fine:

![Screenshot 2024-08-30 at 18 47 31](https://github.com/user-attachments/assets/236308e4-df30-41e3-a963-b1cca4b6bb37)

**Reply by mbrandonw** [reply]
Which beta are you using? It works fine for me in Xcode 16 beta 6.

**Reply by JanC** [reply]
Also Xcode 16 beta 6 🤷‍♂️

Anyway, it seems like an issue related to my local setup and probably not TCA. I'll try to debug further and will close this
